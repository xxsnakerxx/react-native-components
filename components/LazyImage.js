import React from 'react';
import PropTypes from 'prop-types';

import {
  StyleSheet,
  View,
  ViewPropTypes,
  Animated,
  Image,
  Platform,
} from 'react-native';

const sourceIsSame = (src1, src2) => JSON.stringify(src1) === JSON.stringify(src2);

export default class LazyImage extends React.Component {
  static cache = [];

  static propTypes = {
    containerStyle: ViewPropTypes.style,
    style: Image.propTypes.style, // eslint-disable-line react/forbid-foreign-prop-types
    source: Image.propTypes.source, // eslint-disable-line react/forbid-foreign-prop-types
    placeholder: PropTypes.node,
    autoLoad: PropTypes.bool,
    resizeMode: PropTypes.string,
    onLoad: PropTypes.func,
    onError: PropTypes.func,
  };

  static defaultProps = {
    containerStyle: null,
    style: null,
    source: null,
    placeholder: null,
    autoLoad: true,
    resizeMode: 'cover',
    onLoad: () => {},
    onError: () => {},
  }

  constructor(props) {
    super(props);

    this.state = {
      loaded: false,
      src: null,
      opacity: new Animated.Value(0),
    };
  }

  componentDidMount() {
    const {
      autoLoad,
    } = this.props;

    if (autoLoad) {
      this.load();
    }
  }

  componentDidUpdate(prevProps) {
    const {
      source,
      autoLoad,
    } = this.props;

    if (!sourceIsSame(prevProps.source, source)) {
      this.reset();
    }

    if (autoLoad) {
      this.load();
    }
  }

  _onLoadError = () => {
    const {
      onError,
    } = this.props;

    onError(this);
  };

  _onLoad = () => {
    const {
      src,
      opacity,
    } = this.state;

    const {
      onLoad,
    } = this.props;

    if (!src) return;

    this.setState({
      loaded: true,
    });

    const loadedFromCache = LazyImage.cache.includes(JSON.stringify(src));

    if (!loadedFromCache) {
      LazyImage.cache.push(JSON.stringify(src));
    }

    requestAnimationFrame(() => {
      Animated.timing(opacity,
        {
          toValue: 1,
          duration: 300,
          useNativeDriver: Platform.OS === 'ios',
        },
      ).start(() => {
        onLoad(this);
      });
    });
  };

  load() {
    const {
      source,
    } = this.props;

    const {
      loaded,
      src,
    } = this.state;

    if (!loaded && !sourceIsSame(source, src)) {
      this.setState({
        src: source,
      });
    }
  }

  isLoaded() {
    const {
      loaded,
    } = this.state;

    return loaded;
  }

  reset(cb) {
    const {
      opacity,
    } = this.state;

    opacity.setValue(0);

    this.setState({
      loaded: false,
      src: null,
    }, cb);
  }

  render() {
    const {
      containerStyle,
      style,
      placeholder,
      resizeMode,
    } = this.props;

    const {
      src,
      opacity,
    } = this.state;

    return (
      <View
        style={[
          containerStyle,
          styles.container,
        ]}
      >
        {placeholder}
        <Animated.Image
          source={src || null}
          style={[
            styles.image,
            style,
            {
              opacity,
            },
          ]}
          resizeMode={resizeMode}
          onError={this._onLoadError}
          onLoad={this._onLoad}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    overflow: 'hidden',
  },
  image: { ...StyleSheet.absoluteFillObject },
});
