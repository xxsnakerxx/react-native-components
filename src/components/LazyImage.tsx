import React, {ReactElement} from 'react';

import {
  StyleSheet,
  View,
  Animated,
  ImageSourcePropType,
  ImageProps,
  StyleProp,
  ViewStyle,
  NativeSyntheticEvent,
  ImageLoadEventData,
} from 'react-native';

const cache: string[] = [];

const sourceIsCached = (source: ImageSourcePropType): boolean => {
  if (!source) {
    return false;
  }

  if (typeof source !== 'number' && !Array.isArray(source)) {
    return !!source.uri && cache.includes(source.uri);
  }

  // @ts-ignore
  return cache.includes(source);
};

const cacheSource = (source: ImageSourcePropType) => {
  if (!source) {
    return;
  }

  if (typeof source !== 'number' && !Array.isArray(source) && source.uri) {
    cache.push(source.uri);
  }

  // @ts-ignore
  cache.push(source);
};

interface LazyImageProps {
  placeholder?: ReactElement;
  containerStyle?: StyleProp<ViewStyle>;
}

export type Props = LazyImageProps & ImageProps;

interface State {
  opacity: Animated.Value;
}

export default class LazyImage extends React.PureComponent<Props, State> {
  static defaultProps = {
    placeholder: null,
    resizeMode: 'cover',
    onLoad: () => {},
    onLoadStart: () => {},
  };

  constructor(props: Props) {
    super(props);

    this.state = {
      opacity: new Animated.Value(+sourceIsCached(props.source)),
    };
  }

  UNSAFE_componentWillReceiveProps(nextProps: Props) {
    const {opacity} = this.state;

    if (!sourceIsCached(nextProps.source)) {
      opacity.setValue(0);
    }
  }

  _onLoadStart = () => {
    const {opacity} = this.state;

    const {onLoadStart, source} = this.props;

    opacity.setValue(+sourceIsCached(source));

    onLoadStart!();
  };

  _onLoad = (e: NativeSyntheticEvent<ImageLoadEventData>) => {
    const {opacity} = this.state;

    const {onLoad, source} = this.props;

    const {width, height} = e.nativeEvent.source;

    if (width <= 1 || height <= 1) {
      return;
    }

    if (sourceIsCached(source)) {
      opacity.setValue(1);
    } else {
      cacheSource(source);

      requestAnimationFrame(() => {
        Animated.timing(opacity, {
          toValue: 1,
          duration: 300,
          useNativeDriver: true,
        }).start(() => {
          onLoad!(e);
        });
      });
    }
  };

  render() {
    const {containerStyle, placeholder, ...imageProps} = this.props;

    const {style} = imageProps;

    const {opacity} = this.state;

    return (
      <View testID="LazyImage" style={[containerStyle, styles.container]}>
        {placeholder}
        <Animated.Image
          {...imageProps}
          testID="LazyImage-Image"
          onLoad={this._onLoad}
          onLoadStart={this._onLoadStart}
          style={[
            styles.image,
            style,
            {
              opacity,
            },
          ]}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    overflow: 'hidden',
  },
  image: {
    ...StyleSheet.absoluteFillObject,
    backgroundColor: 'white',
  },
});
