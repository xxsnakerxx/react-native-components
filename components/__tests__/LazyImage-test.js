import {
  Text,
} from 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import LazyImage from '../LazyImage';

const containerStyle = { width: 40, height: 40 };
const imageStyle = { borderRadius: 20 };
const onLoad = jest.fn();
const source = { uri: 'https://pic.com/1' };

describe('LazyImage', () => {
  it('renders correctly', () => {
    const component = renderer.create(
      <LazyImage
        containerStyle={containerStyle}
        style={imageStyle}
        source={source}
        placeholder={(
          <Text>
            Placeholder
          </Text>
        )}
      />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders correctly when loaded', () => {
    jest.useFakeTimers();

    const component = renderer.create(
      <LazyImage
        containerStyle={containerStyle}
        style={imageStyle}
        source={source}
        placeholder={(
          <Text>
            Placeholder
          </Text>
        )}
        onLoad={onLoad}
      />,
    );

    const instance = component.getInstance();

    instance._onLoad();
    jest.runAllTimers();

    expect(component.toJSON()).toMatchSnapshot();
    expect(onLoad).toBeCalled();
    expect(instance.isLoaded()).toBeTruthy();
    expect(LazyImage.cache.includes(JSON.stringify(source))).toBeTruthy();
  });

  it('renders correctly when reseted', () => {
    jest.useFakeTimers();

    const component = renderer.create(
      <LazyImage
        containerStyle={containerStyle}
        style={imageStyle}
        source={source}
        placeholder={(
          <Text>
            Placeholder
          </Text>
        )}
        onLoad={onLoad}
      />,
    );

    const instance = component.getInstance();

    instance._onLoad();
    jest.runAllTimers();
    instance.reset(() => {
      expect(component.toJSON()).toMatchSnapshot();
    });
  });
});
