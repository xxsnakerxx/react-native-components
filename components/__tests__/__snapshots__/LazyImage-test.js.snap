// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LazyImage renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "height": 40,
        "width": 40,
      },
      Object {
        "overflow": "hidden",
      },
    ]
  }
>
  <Text>
    Placeholder
  </Text>
  <Image
    onError={[Function]}
    onLoad={[Function]}
    resizeMode="cover"
    source={
      Object {
        "uri": "https://pic.com/1",
      }
    }
    style={
      Object {
        "borderRadius": 20,
        "bottom": 0,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</View>
`;

exports[`LazyImage renders correctly when loaded 1`] = `
<View
  style={
    Array [
      Object {
        "height": 40,
        "width": 40,
      },
      Object {
        "overflow": "hidden",
      },
    ]
  }
>
  <Text>
    Placeholder
  </Text>
  <Image
    onError={[Function]}
    onLoad={[Function]}
    resizeMode="cover"
    source={
      Object {
        "uri": "https://pic.com/1",
      }
    }
    style={
      Object {
        "borderRadius": 20,
        "bottom": 0,
        "left": 0,
        "opacity": 1,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</View>
`;

exports[`LazyImage renders correctly when reseted 1`] = `
<View
  style={
    Array [
      Object {
        "height": 40,
        "width": 40,
      },
      Object {
        "overflow": "hidden",
      },
    ]
  }
>
  <Text>
    Placeholder
  </Text>
  <Image
    onError={[Function]}
    onLoad={[Function]}
    resizeMode="cover"
    source={null}
    style={
      Object {
        "borderRadius": 20,
        "bottom": 0,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</View>
`;
